scripts/update_license_headers.py

```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env python3
+
+import os
+import re
+import sys
+
+def update_license_header(file_path):
+    with open(file_path, 'r') as file:
+        content = file.read()
+
+    # patterns
+    apache_header_pattern = re.compile(
+        r'(/\*|#|//)\s*Licensed under the Apache License, Version 2\.0.*?'
+        r'limitations under the License\.',
+        re.DOTALL
+    )
+    spdx_header = "// SPDX-License-Identifier: Apache-2.0\n"
+
+    # Check if the file already has the SPDX header
+    if content.startswith(spdx_header):
+        print(f"Skipping {file_path}: SPDX header already present")
+        return False
+
+    # Replace Apache header with SPDX
+    new_content, count = apache_header_pattern.subn(spdx_header, content, count=1)
+
+    if count > 0:
+        with open(file_path, 'w') as file:
+            file.write(new_content)
+        print(f"Updated {file_path}")
+        return True
+    else:
+        print(f"Skipping {file_path}: No Apache header found")
+        return False
+
+def process_directory(directory):
+    updated_files = 0
+    for root, _, files in os.walk(directory):
+        for file in files:
+            if file.endswith(('.go')):
+                file_path = os.path.join(root, file)
+                if update_license_header(file_path):
+                    updated_files += 1
+    return updated_files
+
+def main():
+    if len(sys.argv) != 2:
+        print("Usage: python update_license_headers.py <directory>")
+        sys.exit(1)
+
+    directory = sys.argv[1]
+    if not os.path.isdir(directory):
```

Comment:
it looks like this only works on a single directory? What is your suggested way to run it on the whole code base?

Reply:
another alternative is to work on a single file at a time and then we can run it like this:
```
find . -name '*.go' | xargs scripts/update_license_headers.py
```

Reply:
Thanks for the feedback @yurishkuro. 

We can extend the script to this 

```
def process_path(path):
    if os.path.isfile(path) and path.endswith('.go'):
        return 1 if update_license_header(path) else 0
    elif os.path.isdir(path):
        return process_directory(path)
    else:
        print(f"Skipping {path}: Not a .go file or directory")
        return 0

def process_directory(directory):
    updated_files = 0
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.go'):
                file_path = os.path.join(root, file)
                if update_license_header(file_path):
                    updated_files += 1
    return updated_files

def main():
    if len(sys.argv) < 2:
        print("Usage: python update_license_headers.py <file_or_directory> [<file_or_directory> ...]")
        sys.exit(1)

    total_updated = 0
    for path in sys.argv[1:]:
        total_updated += process_path(path)

    print(f"Total files updated: {total_updated}")
```
- It can process single or multiple .go files directly
- Handle entire directories, searching for .go files within
- `find` tool also will work

For eg.

`python scripts/update_license_headers.py path/to/the/directory
`

`find . -name '*.go' | xargs python scripts/update_license_headers.py`

Let me know if you want me to go ahead with the changes.

Reply:
I would rather make the script simpler than more complex. Why build directory descent if I can achieve that just as easily with `find`?

Reply:
You are totally right. I got a bit carried away with trying to make it do everything.

We can update the script to this 

```
#!/usr/bin/env python3
import re
import sys

def update_license_header(file_path):
    with open(file_path, 'r') as file:
        content = file.read()

    apache_header_pattern = re.compile(
        r'(/\*|#|//)\s*Licensed under the Apache License, Version 2\.0.*?'
        r'limitations under the License\.',
        re.DOTALL
    )
    spdx_header = "// SPDX-License-Identifier: Apache-2.0\n"

    if content.startswith(spdx_header):
        print(f"Skipping {file_path}: SPDX header already present")
        return False

    new_content, count = apache_header_pattern.subn(spdx_header, content, count=1)

    if count > 0:
        with open(file_path, 'w') as file:
            file.write(new_content)
        print(f"Updated {file_path}")
        return True
    else:
        print(f"Skipping {file_path}: No Apache header found")
        return False

def main():
    if len(sys.argv) < 2:
        print("Usage: python update_license_headers.py <file> [<file> ...]")
        sys.exit(1)

    total_updated = sum(update_license_header(file) for file in sys.argv[1:])
    print(f"Total files updated: {total_updated}")

if __name__ == "__main__":
    main()
```
And we can lean on `find` for the heavy lifting.

Reply:
just update the PR, no need to post code snippets

Reply:
updated. thanks